src/librbd/managed_lock/AcquireRequest.cc

```diff
@@ -0,0 +1,332 @@
+    create_rados_ack_callback<klass, &klass::handle_get_watchers>(this);
+  m_out_bl.clear();
+  int r = m_ioctx.aio_operate(m_oid, rados_completion, &op, &m_out_bl);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template <typename L>
+void AcquireRequest<L>::handle_get_watchers(int r) {
+  ldout(m_cct, 10) << __func__ << ": r=" << r << dendl;
+
+  if (r == 0) {
+    r = m_watchers_ret_val;
+  }
+  save_result(r);
+  if (r < 0) {
+    lderr(m_cct) << "failed to retrieve watchers: " << cpp_strerror(r) << dendl;
+    finish();
+    return;
+  }
+
+  for (auto &watcher : m_watchers) {
+    if ((strncmp(m_locker_address.c_str(),
+                 watcher.addr, sizeof(watcher.addr)) == 0) &&
+        (m_locker_handle == watcher.cookie)) {
+      ldout(m_cct, 10) << "lock owner is still alive" << dendl;
+
+      save_result(-EAGAIN);
+      finish();
+      return;
+    }
+  }
+
+  send_blacklist();
+}
+
+template <typename L>
+void AcquireRequest<L>::send_blacklist() {
+  if (!m_cct->_conf->rbd_blacklist_on_break_lock) {
+    send_break_lock();
+    return;
+  }
+  ldout(m_cct, 10) << __func__ << dendl;
+
+  // TODO: need async version of RadosClient::blacklist_add
+  using klass = AcquireRequest;
+  Context *ctx = create_context_callback<klass, &klass::handle_blacklist>(
+    this);
+  m_watcher->work_queue()->queue(
+      new C_BlacklistClient(m_ioctx, m_locker_address, ctx), 0);
+}
+template <typename L>
+void AcquireRequest<L>::handle_blacklist(int r) {
+  ldout(m_cct, 10) << __func__ << ": r=" << r << dendl;
+
+  save_result(r);
+  if (r < 0) {
+    lderr(m_cct) << "failed to blacklist lock owner: " << cpp_strerror(r)
+                 << dendl;
+    finish();
+    return;
+  }
+  send_break_lock();
+}
+
+template <typename L>
+void AcquireRequest<L>::send_break_lock() {
+  ldout(m_cct, 10) << __func__ << dendl;
+
+  librados::ObjectWriteOperation op;
+  rados::cls::lock::break_lock(&op, RBD_LOCK_NAME, m_locker_cookie,
+                               m_locker_entity);
+
+  using klass = AcquireRequest;
+  librados::AioCompletion *rados_completion =
+    create_rados_safe_callback<klass, &klass::handle_break_lock>(this);
+  int r = m_ioctx.aio_operate(m_oid, rados_completion, &op);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template <typename L>
+void AcquireRequest<L>::handle_break_lock(int r) {
+  ldout(m_cct, 10) << __func__ << ": r=" << r << dendl;
+
+  if (r == -ENOENT) {
+    r = 0;
+  } else if (r < 0) {
+    lderr(m_cct) << "failed to break lock: " << cpp_strerror(r) << dendl;
+    save_result(r);
+    finish();
+    return;
+  }
+
+  send_lock();
+}
+
+template <typename L>
+void AcquireRequest<L>::finish() {
+  m_on_finish->complete(m_error_result);
+  delete this;
+}
+
+} // namespace managed_lock
+} // namespace librbd
+
+#include "librbd/managed_lock/LockWatcher.h"
+template class librbd::managed_lock::AcquireRequest<
+                                            librbd::managed_lock::LockWatcher>;
```

Comment:
Minor: I would just recommend a single indent when you run into overflow cases instead of as far indented as possible. General comment that applies everywhere.